src/coordinator/storage/local/storage.go

```diff
@@ -90,9 +143,54 @@ func (s *localStorage) FetchTags(ctx context.Context, query *storage.FetchQuery,
 		return nil, err
 	}
 
-	opts := storage.FetchOptionsToM3Options(options, query)
+	var (
+		opts       = storage.FetchOptionsToM3Options(options, query)
+		namespaces = s.clusters.ClusterNamespaces()
+		now        = time.Now()
+		fetches    = 0
+		result     multiFetchTagsResult
+		wg         sync.WaitGroup
+	)
+	for _, namespace := range namespaces {
+		namespace := namespace // Capture var
+
+		clusterStart := now.Add(-1 * namespace.Attributes().Retention)
+
+		// Only include if cluster can completely fulfill the range
+		if clusterStart.After(query.Start) {
+			continue
+		}
+
+		fetches++
+
+		wg.Add(1)
+		go func() {
+			result.add(s.fetchTags(namespace, m3query, opts))
```

Comment:
same as above: do we want to exit on first error ? If yes, then probably use ExecuteParallel, if not, then probably add a comment since we shortcircuit on first error in most cases.

Reply:
I'd have to create temporary objects (i.e. allocation per fanout thing) to pass to ExecuteParallel which is costly, that's why I use a single static waitgroup and only if there are errors does it allocate a per fanout element.